cardinal/world.go

```diff
@@ -0,0 +1,235 @@
+package cardinal
+
+import (
+	"context"
+	"errors"
+	"time"
+
+	"github.com/rs/zerolog/log"
+	"pkg.world.dev/world-engine/cardinal/ecs"
+	"pkg.world.dev/world-engine/cardinal/ecs/inmem"
+	"pkg.world.dev/world-engine/cardinal/ecs/storage"
+	"pkg.world.dev/world-engine/cardinal/ecs/transaction"
+	"pkg.world.dev/world-engine/cardinal/server"
+	"pkg.world.dev/world-engine/cardinal/shard"
+)
+
+type World struct {
+	impl          *ecs.World
+	loopInterval  time.Duration
+	serverOptions []server.Option
+}
+
+// System is a function that process the transaction in the given transaction queue.
+// Systems are automatically called during a world tick, and they must be registered
+// with a world using AddSystem or AddSystems.
+type System func(*World, *TransactionQueue) error
+
+type (
+	// EntityID represents a single entity in the World. An EntityID is tied to
+	// one or more components.
+	EntityID = storage.EntityID
+	TxHash   = transaction.TxHash
+)
+
+// NewWorld creates a new World object using Redis as the storage layer.
+func NewWorld(addr, password string, opts ...WorldOption) (*World, error) {
+	ecsOptions, serverOptions, cardinalOptions := separateOptions(opts)
+	log.Log().Msg("Running in normal mode, using external Redis")
+	if addr == "" {
+		return nil, errors.New("redis address is required")
+	}
+	if password == "" {
+		log.Log().Msg("Redis password is not set, make sure to set up redis with password in prod")
+	}
+
+	rs := storage.NewRedisStorage(storage.Options{
+		Addr:     addr,
+		Password: password, // make sure to set this in prod
+		DB:       0,        // use default DB
+	}, "world")
+	worldStorage := storage.NewWorldStorage(&rs)
+	ecsWorld, err := ecs.NewWorld(worldStorage, ecsOptions...)
+	if err != nil {
+		return nil, err
+	}
+
+	world := &World{
+		impl:          ecsWorld,
+		serverOptions: serverOptions,
+	}
+	for _, opt := range cardinalOptions {
+		opt(world)
+	}
+
+	return world, nil
+}
+
+// NewMockWorld creates a World that uses an in-memory redis DB as the storage layer.
+// This is only suitable for local development.
+func NewMockWorld(opts ...WorldOption) (*World, error) {
+	ecsOptions, serverOptions, cardinalOptions := separateOptions(opts)
+	world := &World{
+		impl:          inmem.NewECSWorld(ecsOptions...),
+		serverOptions: serverOptions,
+	}
+	for _, opt := range cardinalOptions {
+		opt(world)
+	}
+	return world, nil
+}
+
+func toITransactionType(ins []AnyTransaction) []transaction.ITransaction {
```

Comment:
this might be able to go to utils.go instead?

Reply:
done